### PR TITLE
Add tracking for "Product Added" Segment event

### LINKED ISSUE #360: Add tracking to "Add to cart" action
# Add tracking to "Add to cart" action

### Summary
When a user click the "Add to cart" button, a segment tracking action needs to be executed.

### PR DESCRIPTION
**Pull request by willopez**
Resolves #360 
Impact: minor
Type: feature

## Summary
Add tracking for "Product Added" Segment event.

## Testing
1. Click on a product to enter the product's details page
2. Verify that a "Product Viewed" event was recorded in the Segment dashboard
4. Increment a product's quantity to an arbitrary number
3. Click the "Add to cart" button
4. Verify a "Product Added" was recorded in the Segment dashboard. The event should include the correct quantity and cartId.
5. Login
6. Repeat steps 1-4,

### CODE REVIEW

File: docs/tracking-events.md [hunk 1]
```diff
@@ -238,12 +238,16 @@ class ProductDetailPage extends Component {
...
-  @trackProductViewed()
   selectVariant(variant, optionId) {
     // Do something with selected variant / option
+
+    this.trackAction({ variant, optionId, action: "Product Viewed" });
   }
 
+  // Expects the prop `product`, and an object with the following keys `variant`, `optionId`, and `action` as a function arg. The `product` prop is provided by the `ProductDetailPage` component.
+  @trackProduct()
+  trackAction(functionArgs) {}
```
Comment (mikemurray) [inline docs/tracking-events.md]: See below comment about maybe not needing the `functionArgs` param. If you don't need to use it then it is best not to include it.

File: src/components/ProductDetail/ProductDetail.js [hunk 2]
```diff
@@ -88,6 +89,9 @@ class ProductDetail extends Component {
     });
   }
 
+  @trackProduct()
+  trackAction(functionArgs) {} // eslint-disable-line no-unused-vars
```
Comment (mikemurray) [inline src/components/ProductDetail/ProductDetail.js]: Is functions args needed, or can it be removed?

File: src/components/ProductDetail/ProductDetail.js [hunk 3]
```diff
@@ -80,6 +79,8 @@ class ProductDetail extends Component {
       selectOptionId = variant.options[0]._id;
     }
 
+    this.trackAction({ variant, optionId, action: "Product Viewed" });
```
Comment (mikemurray) [inline src/components/ProductDetail/ProductDetail.js]: We should make `Product Viewed` a constant and export from either `lib/tracking/constants.js` or from `lib/tracking/trackProduct.js`.

File: src/components/ProductDetail/ProductDetail.js [hunk 4]
```diff
@@ -140,6 +144,23 @@ class ProductDetail extends Component {
...
+        const { cart } = data.createCart || data.addCartItems;
+
+        this.trackAction({
+          variant: {
+            ...selectedVariant,
+            cart_id: cart._id, // eslint-disable-line camelcase
+            quantity
+          },
+          optionId: selectedOption ? selectedOption._id : null,
+          action: "Product Added"
```
Comment (mikemurray) [inline src/components/ProductDetail/ProductDetail.js]: We should make `Product Added` a constant and export from either `lib/tracking/constants.js` or from `lib/tracking/trackProduct.js`.

File: src/containers/cart/withCart.js [hunk 5]
```diff
@@ -430,9 +430,9 @@ export default function withCart(Component) {
                 {(mutationFunction) => (
                   <Component
                     {...this.props}
-                    addItemsToCart={async (items) => {
-                      await this.handleAddItemsToCart(mutationFunction, { items }, !cart);
-                    }}
+                    addItemsToCart={async (items) => (
+                      this.handleAddItemsToCart(mutationFunction, { items }, !cart)
```
Comment (mikemurray) [inline src/containers/cart/withCart.js]: If we're no longer awaiting the result of `handleAddItemsToCart` then `async` should also be removed

File: docs/tracking-events.md [hunk 6]
```diff
@@ -238,12 +238,16 @@ class ProductDetailPage extends Component {
     this.selectVariant(product.variants[0]);
   }
 
-  // expects the prop `product`, with `variant` and `optionId` as function params
-  @trackProductViewed()
   selectVariant(variant, optionId) {
     // Do something with selected variant / option
+
+    this.trackAction({ variant, optionId, action: "Product Viewed" });
```
Comment (mikemurray) [inline docs/tracking-events.md]: See below comment about making `Product Viewed` a constant.

File: docs/tracking-events.md [hunk 7]
```diff
@@ -238,12 +238,16 @@ class ProductDetailPage extends Component {
...
   }
 
+  // Expects the prop `product`, and an object with the following keys `variant`, `optionId`, and `action` as a function arg. The `product` prop is provided by the `ProductDetailPage` component.
```
Comment (mikemurray) [inline docs/tracking-events.md]: Break comment into two lines to avoid wrapping.

File: src/lib/tracking/trackProduct.js [hunk 8]
```diff
@@ -3,36 +3,33 @@ import getProductTrackingData from "./utils/getProductTrackingData";
...
- * @name trackProductViewed
+ * trackProduct HOC tracks a product
+ * @name trackProduct
  * @param {Object} options options to supply to tracking HOC
  * @returns {React.Component} - component
  */
 export default (options) => (
   track(({ product, router }, state, functionArgs) => {
     let data = {};
+    const { variant, optionId, action } = functionArgs[0];
```
Comment (mikemurray) [inline src/lib/tracking/trackProduct.js]: There's a circle CI error complaining about `functionArgs` being `undefined`.

File: src/containers/cart/withCart.js (hunk 5 above)
Comment (willopez) [reply src/containers/cart/withCart.js]: I removed the `async` not sure why it's not reflected in the comment above.

**Comment by willopez** [comment]
@mikemurray ready for another review

**Pull request merged**